File: src/main/java/pl/com/tegess/controller/login/LoginController.java
package pl.com.tegess.controller.login;

import org.bson.types.ObjectId;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.servlet.view.RedirectView;
import pl.com.tegess.controller.login.facebook.FacebookLoginHelper;
import pl.com.tegess.controller.login.facebook.FacebookUtils;
import pl.com.tegess.controller.login.request.facebook.FacebookTokenResponse;
import pl.com.tegess.controller.login.request.facebook.FacebookUserData;
import pl.com.tegess.controller.login.request.facebook.FacebookValidateTokenResponse;
import pl.com.tegess.domain.application.Application;
import pl.com.tegess.domain.application.ApplicationRepository;
import pl.com.tegess.domain.user.User;
import pl.com.tegess.domain.user.service.UserService;

@RestController
@Component
public class LoginController {

    @Autowired
    ApplicationRepository repository;
    @Autowired
    private UserService userService;
    @Autowired
    private TokenManager tokenManager;
    @Autowired
    private FacebookLoginHelper facebookLoginHelper;

    private RestTemplate restTemplate = new RestTemplate();

    @RequestMapping(value = "api/login/facebook", method = RequestMethod.GET)
    public RedirectView loginByFacebook(@RequestParam String appId) {

        Application application = repository.findOne(new ObjectId(appId));

        RedirectView view = new RedirectView();
        view.setUrl(FacebookUtils.prepareCodeRequest(application));

        return view;
    }

    @RequestMapping(value = "api/logged/facebook", method = RequestMethod.GET)
    public RedirectView loggedByFacebook(
            @RequestParam String appId,
            @RequestParam String code) throws Exception {

        Application application = repository.findOne(new ObjectId(appId));

        //Firstly we need obtain token from FB.
        FacebookTokenResponse facebookTokenResponse = facebookLoginHelper.getFacebookTokenResponse(code, application);

        //Next, we need get our application token
        String applicationToken = facebookLoginHelper.getFacebookApplicationToken(application);

        //Next, we need validate this token
        FacebookValidateTokenResponse facebookValidateTokenResponse =
                facebookLoginHelper.getFacebookValidateTokenResponse(application, facebookTokenResponse, applicationToken);

        System.out.println("User id = " + facebookValidateTokenResponse.getData().getUser_id());

        //Get user data
        String userInfoRequestURI =
                FacebookUtils.prepareUserInfoRequest(facebookValidateTokenResponse.getData().getUser_id());


        HttpEntity<Object> objectHttpEntity = FacebookUtils.prepareHeadersForFacebookAuthorization(facebookTokenResponse);

        ResponseEntity<FacebookUserData> responseEntity =
                restTemplate.exchange(userInfoRequestURI, HttpMethod.GET, objectHttpEntity, FacebookUserData.class);

        //Create user and add it to db
        User user = userService.createUser(appId, responseEntity.getBody());

        //Generate JWTToken and redirect to Client App.
        String token = tokenManager.generateJWTTokenForUser(user, application);

        String url = application.getRedirectURI() + "?token=" + token;

        RedirectView redirectView = new RedirectView();
        redirectView.setUrl(url);

        return redirectView;
    }
}

File: src/main/scala/com/tegess/controller/login/LoginController.scala
package com.tegess.controller.login

import java.util
import java.util.Locale

import com.avsystem.commons._
import com.tegess.controller._
import com.tegess.controller.login.FacebookHelper.{FacebookApplicationTokenResponse, FacebookTokenResponse, FacebookUserData, FacebookValidateTokenResponse}
import com.tegess.controller.login.LoginController._
import com.tegess.domain.user.{FacebookLogin, User}
import com.tegess.persistance.service.application.ApplicationService
import com.tegess.persistance.service.event.LoginEventService
import com.tegess.persistance.service.user.UserService
import org.bson.types.ObjectId
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpMethod
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Component
import org.springframework.util.Base64Utils
import org.springframework.web.bind
import org.springframework.web.bind.annotation._
import org.springframework.web.client.RestTemplate
import org.springframework.web.servlet.view.RedirectView

import scala.util.{Failure, Success, Try}

@bind.annotation.RestController
@Component
class LoginController {

  @Autowired var applicationService: ApplicationService = _
  @Autowired var userService: UserService = _
  @Autowired var loginEventService: LoginEventService = _
  @Autowired var restTemplate: RestTemplate = _
  @Autowired var passwordEncoder:PasswordEncoder = _

  @RequestMapping(value = Array("api/login/credentials"), method = Array(RequestMethod.GET))
  def loginByCredentials(@RequestParam appId: String, @RequestHeader(value = "Authorization") credentials: String): TokenResponse = {
    val result = for {
      loginPassword <- decodeUserFromCredentials(credentials)
      applicationId <- Try(new ObjectId(appId))
      user <- Try(userService.findOne(applicationId, loginPassword.login).get)
      password <- Try(user.password.get)
      if passwordEncoder.matches(loginPassword.password, password)
      application <- Try(applicationService.findOne(applicationId).get)
      token = TokenGenerator.generateJWTTokenForUser(application, user)
    } yield {
      loginEventService.publishLoginEvent(application, user)
      TokenResponse(token)
    }
    result match {
      case Success(token) => token
      case Failure(e) => throw new IllegalArgumentException("Wrong parameters - ", e)
    }
  }

  @RequestMapping(value=Array("api/login/facebook"), method=Array(RequestMethod.GET))
  def loginByFacebook(@RequestParam appId: String): RedirectView = {
    val response = for {
      applicationId <- Try(new ObjectId(appId))
      application <- Try(applicationService.findOne(applicationId).get)
      fbLogin <- Try(application.fbLogin.get)
    } yield new RedirectView().setup { _.setUrl(FacebookHelper.prepareCodeRequest(applicationId, fbLogin)) }

    response match {
      case Success(r) => r
      case Failure(e) => throw new IllegalArgumentException("Something went wrong - ", e)
    }
  }

  @RequestMapping(value=Array("api/logged/facebook"), method=Array(RequestMethod.GET))
  def loggedByFacebook(@RequestParam appId: String, @RequestParam code: String) = {

    val result = for {
      applicationId <- Try(new ObjectId(appId))
      application <- applicationService.findOne(applicationId).toTry
      fbLogin <- application.fbLogin.toTry
      tokenResponse <- Try(restTemplate.getForObject(
        FacebookHelper.prepareTokenRequest(applicationId, fbLogin, code), classOf[FacebookTokenResponse]))
      applicationTokenResponse <- Try(restTemplate.getForObject(
        FacebookHelper.prepareApplicationTokenRequest(fbLogin), classOf[FacebookApplicationTokenResponse]))
      validateTokenResponse <- Try(restTemplate.getForObject(
        FacebookHelper.prepareValidateTokenRequest(tokenResponse, applicationTokenResponse), classOf[FacebookValidateTokenResponse]))
      if fbLogin.id == validateTokenResponse.data.app_id
      userData <- Try(restTemplate.exchange(
        FacebookHelper.prepareUserInfoRequest(validateTokenResponse.data.user_id),
        HttpMethod.GET,
        FacebookHelper.prepareHeadersForFacebookAuthorization(tokenResponse),
        classOf[FacebookUserData]).getBody)
      user <- Try(User(userData.name,
        applicationId, List(FacebookLogin), None, Some(userData.email),
        Some(userData.picture.data.url), Some(userData.gender), Some(new Locale(userData.locale)), List()))
      token <- Try(TokenGenerator.generateJWTTokenForUser(application, user))
      url <- Try(application.redirectURL.get + s"?token=$token")
    } yield {
      userService.save(user)
      loginEventService.publishLoginEvent(application, user)
      new RedirectView().setup(_.setUrl(url))
    }

    result match {
      case Success(r) => r
      case Failure(e) => throw new IllegalArgumentException("Something went wrong -", e)
    }
  }

}
object LoginController {

  case class LoginPassword(login: String, password: String)
  case class TokenResponse(token: String)

  def decodeUserFromCredentials(credentials: String): Try[LoginPassword] = {
    for {
      decodedCredentials <- Try(new String(Base64Utils.decodeFromString(credentials)))
      slittedCredentials = decodedCredentials.split(":")
      login <- Try(slittedCredentials(0))
      password <- Try(slittedCredentials(1))
    } yield LoginPassword(login, password)
  }

}